---
title: "Genomic Donor Selection"
author: "Daniel Anstett"
date: "25/07/2021"
output: 
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Objective
This markdown shows the R code that generates various plots that may inform selection of potential donor sites for assisted migration. We use climate associated SNPs from whole genome sequencing data for the plant scarlet monkeyflower, _Mimulus cardinalis_ from populations across 55 sites, spanning most of the range of this plant. This page is meant to complement the Conservation Genomics Workshop run by Amy Angert and Daniel Anstett on July 29 & 30th 2021. 


## 1. Import libraries
Import the following libraries that are required:

```{r message=FALSE}
library(tidyverse)
library(cowplot)
library(sf)
library(tmap)
library(rnaturalearth)
library(rnaturalearthdata)
#devtools::install_github("ropensci/rnaturalearthhires") # how to install rnaturalearthdata
library(rnaturalearthhires)
library(rgeos)
library(geodist)
library(RColorBrewer)
library(VennDiagram)
```


## 2. Data Import
Import the data and do some simple calculations and data setup:

```{r message=FALSE}
#Import population data and calculate distance between sites in km
pop_var_raw <- read_csv("Data/paper_ID_site_select.csv")
deg_distance <- pop_var_raw %>% dplyr::select(Long,Lat)
geography <- geodist(deg_distance) 
geography_km <- as.data.frame(geography/1000)

#Import abundance results
freq_1 <- read_csv("Data/freq_1.csv")
freq_2 <- read_csv("Data/freq_2.csv")
freq_5 <- read_csv("Data/freq_5.csv")

#Import presence/abscence results
binary_1 <- read_csv("Data/freq_binary_1.csv")
binary_2 <- read_csv("Data/freq_binary_2.csv")
binary_5 <- read_csv("Data/freq_binary_5.csv")

# Get Proportion of SNPs Present
psp_1 <- as.data.frame(colMeans(binary_1[5:59],na.rm = TRUE))
psp_2 <- as.data.frame(colMeans(binary_2[5:59],na.rm = TRUE))
psp_5 <- as.data.frame(colMeans(binary_5[5:59],na.rm = TRUE))

#Put proportion of SNPs Present into population dataframe 
#Relable the columns according to the climate variable each SNP is associated with
pop_var <- cbind(pop_var_raw,psp_1,psp_2,psp_5)
colnames(pop_var)[6:8] <- c("MAT","MAP","CMD")
```


## 3. Mapping Setup
Set up state boundaries for maps used when making donor selection plots:

```{r message=FALSE}
# California, Oregon & Nevada  Map Setup
states<-ne_states(country=c("canada","united states of america"),returnclass= "sf")
calo <- states %>%
  filter(name_en=="Oregon" | name_en=="California" | name_en=="Nevada")

#Define Coordinate Reference System (CRS) in WGS 1984
EPSG4326<-"+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0" #setup WGS 1984 CRS
```


## 4. Venn diagram of SNP overlap between three climate variables
Visualize the number of SNPs associated with each environmental variable and the number of SNPs that are associated with more than one variable.
For the rest of this markdown please note:  
MAT = Mean annual temperature (Â°C)  
MAP = Mean annual precipitation (mm)  
CMD = Hargreaves climatic moisture deficit (mm)  

```{r message=FALSE}
MAT_set <- pull(binary_1[,1])
MAP_set <- pull(binary_2[,1])
CMD_set <- pull(binary_5[,1])

ven_3 <- list(MAT_set,MAP_set,CMD_set)
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
VD_1<-venn.diagram(x=ven_3,
             category.names = c("MAT" , "MAP" , "CMD"),
             fill = c("yellow", "cyan","magenta"),
             #alpha = c(0.5, 0.5, 0.5),
             main.cex = 5,
             cat.cex = 1.8, cex=1.8,
             fontface = "bold",
             filename = NULL,
)
grid.newpage()
grid.draw(VD_1)

```


## 5. Map the proportion of all SNPs associated climate
Visualize the proportion of all SNPs for each of the 55 sequenced sites. 
We will first do this for mean annual temperature.

```{r message=FALSE}
#Set up sf objects for snp proportions
MAT_points <- pop_var %>% dplyr::select(Long,Lat,MAT)
MAT_sf <- st_as_sf(MAT_points,coords=c("Long","Lat"), crs=EPSG4326)
MAP_points <- pop_var %>% dplyr::select(Long,Lat,MAP)
MAP_sf <- st_as_sf(MAP_points,coords=c("Long","Lat"), crs=EPSG4326)
CMD_points <- pop_var %>% dplyr::select(Long,Lat,CMD)
CMD_sf <- st_as_sf(CMD_points,coords=c("Long","Lat"), crs=EPSG4326)

#Make Map
tmap_mode("view")
MAT_all <- tm_shape(calo)+
  tm_borders()+
  tm_shape(MAT_sf)+
  tm_bubbles(size = 0.1,col="MAT")+ 
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)
```

The proportions per site are given as an interactive map in a similar style to that of ArcGIS. Use the plus an minus buttons to zoom in and out. Click and hold the mouse to grab ahold of the map and move it to whatever region you would like to see. If you would like to change the geographical background from a basic grey map to something resembling google maps, simply hover over the stack icon directly underneath the minus and select either "OpenStreeMap" or "Esri.WorldTopoMap".

Colour ranges for each proportion is given on your left. You can hover over each dot to see the exact proportion. Based on this tool you can graphically identify individual populations that have a certain proportion of the target SNP associated with Mean Annual Temperature.


#### Distribution of the proportion of SNPs associated with increased mean annual temperature

```{r message=FALSE}
MAT_all
```

Our focal plant _M. caridnalis_ occurs across an aridity and temperature gradient from cooler, wetter areas in Oregon, down to warmer, arid regions in Southern California. You can see an increase in temperature adaptive SNPs towards Southern California. However, many sites in central and northern parts of the species range still have SNPs associated with high MAT at greater than 50% of the sites. This suggests the genetic raw material for population adaption to climate change may be present throughout the range.

Now we can produce similar interactive maps for the SNPs associated with the other two climate variables of interest:


```{r message=FALSE}
#MAP
tmap_mode("view")
MAP_all <- tm_shape(calo)+
  tm_borders()+
  tm_shape(MAP_sf)+
  tm_bubbles(size = 0.1,col="MAP")+ 
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)

#CMD
tmap_mode("view")
CMD_all <- tm_shape(calo)+
  tm_borders()+
  tm_shape(CMD_sf)+
  tm_bubbles(size = 0.1,col="CMD")+ 
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)

```

As can be seen in the two panels below the geographical patterns are quite similar for all maps, with greater SNPs associated with lower precipitation and increased moisture deficit (CMD) towards the southern part of the range of _M. cardinalis_. However, do note that this pattern is perhaps strongest for moisture deficit, reflecting the importance of the timing of the onset of terminal drought in this system. Also do note that the central and northern parts of the range still have a considerable proportion of drought associated SNPs, perhaps reflecting the importance of drought throughout the range of the plant.

#### Distribution of the proportion of SNPs associated with decreased mean annual precipitation

```{r message=FALSE}
MAP_all
```

#### Distribution of the proportion of SNPs associated with increased cumulative moisture deficit

```{r message=FALSE}
CMD_all
```

It is also possible to show this graphics in a static printable format by changing from tmap_mode("view") to tmap_mode("plot")

```{r message=FALSE}
#MAT
tmap_mode("plot")
MAT_all <- tm_shape(calo)+
  tm_borders()+
  tm_shape(MAT_sf)+
  tm_bubbles(size = 0.18,col="MAT")+ 
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)
MAT_all
```



## 6. Distance to climate associated SNPs for target site
We can now show how far away each climate associated SNP is from a site of conservation concern. For this example, we will consider site 8, aka Deep Creek, a site in Northern California. We will assume this is the target for possible conservation actions such as assisted migration. 

First we will visualize the location of the site:

```{r message=FALSE}
# Get coordinates for all populations into an sf object
pop_points <- pop_var_raw %>% dplyr::select(Long,Lat) #select relevant data
pop_points_sf <- st_as_sf(pop_points,coords=c("Long","Lat"), crs=EPSG4326)

#Setup sf object for site 8 so it can be plotted separately
p8_only <- pop_var_raw %>% filter(Paper_ID==8) %>% dplyr::select(Long,Lat)
p8_sf <- st_as_sf(p8_only,coords=c("Long","Lat"), crs=EPSG4326)

#Plot population Map with site 8 highlighted
tmap_mode("view")
pop_tmap <- tm_shape(calo)+
  tm_borders()+
  tm_shape(pop_points_sf)+
  tm_bubbles(size = 0.18,alpha=0,border.col="black")+ 
  tm_shape(p8_sf)+
  tm_dots(size=0.3,shape=20,col= "#33FFFF")+
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.005)
pop_tmap
```

We will visualize Deep Creek as a teal circle for the rest of this markdown.


We now calculate the distance from Deep Creek required to find each SNP. 

#### Select shortest distance for each SNP associated with MAT:

```{r message=FALSE}
dist_p8_MAT <- data.frame() #Generate empty data frame
for (i in 1:nrow(binary_1)){
  geography_p8 <- geography_km[,8] #get distance from p8, can be changed to select population of choice
  b1_t <- as.data.frame(t(binary_1[i,5:59])) #get line i from binary matrix
  geo_b1 <- as.data.frame(cbind(geography_p8,b1_t)) #bind columns together
  colnames(geo_b1) <- c("geo","snp")  #rename
  geo_b1 <-  geo_b1 %>% filter(snp==1) #select for presence of "adaptive" snps
  geo_b1 <-  geo_b1 %>% filter(geo==min(geo)) #select for minimum distance
  dist_p8_MAT[i,1] <- geo_b1[1,1] #write out in distance for i snp
}
colnames(dist_p8_MAT) <- "Distance"
```

#### Make graphic showing distance to closest SNP associated with MAT:

```{r message=FALSE, warning=FALSE}
#Make MAT graphic
dist_MAT <- ggplot(data=dist_p8_MAT, aes(dist_p8_MAT$Distance))+
  geom_histogram(binwidth = 20)+
  scale_x_continuous(name="Distance (km)")+
  scale_y_continuous(name="Number of SNPs Associated with MAT")+
  theme_classic()
dist_MAT + theme(
  axis.text.x = element_text(size=14,face="bold"),
  axis.text.y = element_text(size=14,face="bold"),
  axis.title.x = element_text(color="black", size=24, vjust = 0.5, face="bold"),
  axis.title.y = element_text(color="black", size=16,vjust = 2, face="bold",hjust=0.5))
```

Notice that for most SNPs the distance is zero. This means that the wanted SNP is already present in the at risk population. This is a rather positive outcome since it indicates that the population might have enough genetic variability to adapt to increases in temperature. However, there is a proportion of SNPs that are not present. If there is an interest in ensuring this genetic variation arrives at the target site, assisted migration might be required. 

Next we will check the distance to SNPs associated with the other two environmental variables:

#### Set up a dataframe with shortest distance for each SNP associated with MAP and with CMD
```{r message=FALSE}
# MAP
dist_p8_MAP <- data.frame() #Generate empty data frame
for (i in 1:nrow(binary_2)){
  geography_p8 <- geography_km[,8] #get distance from p8
  b2_t <- as.data.frame(t(binary_2[i,5:59])) #get line i from binary matrix
  geo_b2 <- as.data.frame(cbind(geography_p8,b2_t)) #bind columns together
  colnames(geo_b2) <- c("geo","snp")  #rename
  geo_b2 <-  geo_b2 %>% filter(snp==1) #select for presence of "adaptive" snps
  geo_b2 <-  geo_b2 %>% filter(geo==min(geo)) #select for minimum distance
  dist_p8_MAP[i,1] <- geo_b2[1,1] #write out in distance for i snp
}
colnames(dist_p8_MAP) <- "Distance"

# CMD
dist_p8_CMD <- data.frame() #Generate empty data frame
for (i in 1:nrow(binary_5)){
  geography_p8 <- geography_km[,8] #get distance from p8
  b5_t <- as.data.frame(t(binary_5[i,5:59])) #get line i from binary matrix
  geo_b5 <- as.data.frame(cbind(geography_p8,b5_t)) #bind columns together
  colnames(geo_b5) <- c("geo","snp")  #rename
  geo_b5 <-  geo_b5 %>% filter(snp==1) #select for presence of "adaptive" snps
  geo_b5 <-  geo_b5 %>% filter(geo==min(geo)) #select for minimum distance
  dist_p8_CMD[i,1] <- geo_b5[1,1] #write out in distance for i snp
}
colnames(dist_p8_CMD) <- "Distance"

```


#### Then make a graphic showing distance to closest SNP associated with MAP and with CMD:

```{r message=FALSE,warning=FALSE}
#Make MAP graphic
dist_MAP <- ggplot(data=dist_p8_MAP, aes(dist_p8_MAP$Distance))+
  geom_histogram(binwidth = 20)+
  scale_x_continuous(name="Distance (km)")+
  scale_y_continuous(name="Number of SNPs Associated with MAP")+
  theme_classic()
dist_MAP + theme(
  axis.text.x = element_text(size=14,face="bold"),
  axis.text.y = element_text(size=14,face="bold"),
  axis.title.x = element_text(color="black", size=24, vjust = 0.5, face="bold"),
  axis.title.y = element_text(color="black", size=16,vjust = 2, face="bold",hjust=0.5))

#Make CMD graphic
dist_CMD <- ggplot(data=dist_p8_CMD, aes(dist_p8_CMD$Distance))+
  geom_histogram(binwidth = 20)+
  scale_x_continuous(name="Distance (km)")+
  scale_y_continuous(name="Number of SNPs Associated with CMD", limits=c(0,3000))+
  theme_classic()
dist_CMD + theme(
  axis.text.x = element_text(size=14,face="bold"),
  axis.text.y = element_text(size=14,face="bold"),
  axis.title.x = element_text(color="black", size=24, vjust = 0.5, face="bold"),
  axis.title.y = element_text(color="black", size=16,vjust = 2, face="bold",hjust=0.5))
```


Both graphs show a similar pattern to MAT, making a potential conservation decision similar regardless of what association is considered.


### 7. Plot missing SNPs for Deep Creek
As we saw in previous distance graphs, there are a number of SNPs not present at the target site. We can now map the location of only the SNPs not present at Deep Creek. 

#### Set up the sf object for the SNPs associated with MAP that are not present at Deep Creek: 
```{r message=FALSE}
##Setup P8 MAT
binary1_miss_p8 <- binary_1 %>% filter(P8==0) #select only snps not in P8, can be changed to selection population of choice
psp_1_p8 <- as.data.frame(colMeans(binary1_miss_p8[5:59],na.rm = TRUE))
pop_var_p8 <- cbind(pop_var_raw,psp_1_p8)
colnames(pop_var_p8)[6] <- "MAT" 

#Population values for p8
MAT_points_p8 <- pop_var_p8 %>% dplyr::select(Long,Lat,MAT)
MAT_sf_p8 <- st_as_sf(MAT_points_p8,coords=c("Long","Lat"), crs=EPSG4326)
```


#### Make the map for SNPs associated with MAT that are not present at Deep Creek:
```{r message=FALSE}
tmap_mode("view")
miss_p8_MAT <- tm_shape(calo)+
  tm_borders()+
  tm_shape(MAT_sf_p8)+
  tm_bubbles(size = 0.15,col="MAT")+ 
  tm_shape(p8_sf)+
  tm_dots(size=0.3,shape=20,col= "#33FFFF")+
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)
miss_p8_MAT
```

There is one site in the same region as Deep Creek that has > 40% of SNPs not present at Deep Creek. This site may be particularly valuable for assisted migration given their geographical proximity. If even more SNPs are desired there are also locations in the centre and many in the south of the range that have even more SNPs associated with higher temperatures.

Again we can also look at maps for MAP and CMD.

#### Set up the sf object for the SNPs associated with MAP and with CMD that are not present at Deep Creek: 
```{r message=FALSE}
##Setup P8 MAP
binary2_miss_p8 <- binary_2 %>% filter(P8==0) #select only snps not in P8
PSP_2_p8 <- as.data.frame(colMeans(binary2_miss_p8[5:59],na.rm = TRUE))
pop_2var_p8 <- cbind(pop_var_raw,PSP_2_p8)
colnames(pop_2var_p8)[6] <- "MAP" 

#Population values for p8
MAP_points_p8 <- pop_2var_p8 %>% dplyr::select(Long,Lat,MAP)
MAP_sf_p8 <- st_as_sf(MAP_points_p8,coords=c("Long","Lat"), crs=EPSG4326)

##Setup P8 CMD
binary5_miss_p8 <- binary_5 %>% filter(P8==0) #select only snps not in P8
PSP_5_p8 <- as.data.frame(colMeans(binary5_miss_p8[5:59],na.rm = TRUE))
pop_5var_p8 <- cbind(pop_var_raw,PSP_5_p8)
colnames(pop_5var_p8)[6] <- "CMD" 

#Population values for p8
CMD_points_p8 <- pop_5var_p8 %>% dplyr::select(Long,Lat,CMD)
CMD_sf_p8 <- st_as_sf(CMD_points_p8,coords=c("Long","Lat"), crs=EPSG4326)
```


#### Make the map for SNPs associated with MAP that are not present at Deep Creek:
```{r message=FALSE}
tmap_mode("view")
miss_p8_MAP <- tm_shape(calo)+
  tm_borders()+
  tm_shape(MAP_sf_p8)+
  tm_bubbles(size = 0.15,col="MAP")+ 
  tm_shape(p8_sf)+
  tm_dots(size=0.3,shape=20,col= "#33FFFF")+
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)
miss_p8_MAP

```

Again SNPs missing for our target site are present throughout the range. Just over 45% of the variation missing is present at a nearby site still within the northern region. Sites with higher levels of variation are also available in the central and especially southern parts of the range.


#### Make the map for SNPs associated with CMD that are not present at Deep Creek:
```{r message=FALSE}
tmap_mode("view")
miss_p8_CMD <- tm_shape(calo)+
  tm_borders()+
  tm_shape(CMD_sf_p8)+
  tm_bubbles(size = 0.15,col="CMD")+ 
  tm_shape(p8_sf)+
  tm_dots(size=0.3,shape=20,col= "#33FFFF")+
  tm_layout(legend.position = c(0.29, 0.73),legend.title.size = 0.001)
miss_p8_CMD
```

CMD shows a somewhat different pattern and is of particular importance given that it is our best metric of drought since it takes into account both temperature and lack of precipitation. Some of the missing SNPs are present in northern sites, but each population has <40% of the SNPs. To find SNPs with > 50 % of the associated SNPs central populations are required, and for >60 % of the SNPs only southern populations are appropriate donor sites.


#### 8. Conclusion

Overall we provide a multiple of visualization tools that we hope are a useful starting point in search for possible donor sites and conservation strategies, and as a way consider the types of information that can be gleaned from landscape genomics. A full version of the code used to compute these and other similar maps is available in our github repo "conservation_genomics". 




